app/src/main/java/com/example/cactusnotes/SignUpActivity.kt

```diff
@@ -18,15 +37,54 @@ class SignUpActivity : AppCompatActivity() {
 
         binding.signUpButton.setOnClickListener {
             if (isEmailValid() && isUsernameValid() && isPasswordValid()) {
-                Toast.makeText(applicationContext, R.string.successful, Toast.LENGTH_LONG).show()
+                val request = CactusModel(
+                    binding.emailEditText.editText?.text.toString(),
+                    binding.userNameEditText.editText?.text.toString(),
+                    binding.passwordEditText.editText?.text.toString()
+                )
+
+                api.register(request).enqueue(object : Callback<RegisterResponse> {
+                    override fun onResponse(
+                        call: Call<RegisterResponse>,
+                        response: Response<RegisterResponse>
+                    ) {
+                        if (response.isSuccessful) {
+                            Snackbar.make(binding.root, R.string.successful, Snackbar.LENGTH_LONG)
+                                .show()
+                        } else {
+                            val errorBody = response.errorBody()!!.string()
+                            val errorResponse = GsonBuilder()
+                                .create()
+                                .fromJson(errorBody, RegisterErrorResponse::class.java)
```

Comment:
This line could lead to crashes since you are not sure of the structure of the response. You can test if the error response can be deserialized into the given class (RegisterErrorResponse) by wrapping that logic with a try catch block.